README

Hidden Markov Models with R



NEWS

Changes in hmmr version 0.3-0

  o Added many datasets to accompany the book. 
  
Changes in hmmr version 0.2-0
  
  o  Functions hmm and lca fully operational. 
  
First R CMD INSTALL'able version CRAN: 0.1-0 (28-6-2011, Rev #17)


R/hmm.R
lca <- function(data, nclasses, fit=TRUE, family=NULL, verbose=TRUE, ...) {
	
	# univariate numeric data by default treated as gaussian
 	if(is.vector(data)) {
		nm <- deparse(substitute(data))
		form <- as.formula(paste(nm,"~1",sep=""))
		if(is.null(family)) family=gaussian()
		simple <- 0
		data <- data.frame(nm=data)
	} else {
		# univariate factor data treated as multinomial
 		if(is.factor(data)) {
			nm <- deparse(substitute(data))
			form <- as.formula(paste(nm,"~1",sep=""))
			if(is.null(family)) family=multinomial("identity")
			simple <- 1
			data <- data.frame(nm=data)
		} else {
			# multivariate data
			form <- list()
			fam <- list()
			nc <- ifelse(!is.null(dim(data)), dim(data)[2],1)
			simple <- numeric(nc)
			for(i in 1:nc) {
				form[[i]] <- as.formula(paste(names(data)[i],"~1",sep=""))
				fam[[i]] <- switch(class(data[,i]),
					"numeric"=gaussian(),
					"factor"=multinomial("identity"),
					stop("Provide family arguments of data other than 'numeric' or 
						'factor's"))
				simple[i] <- switch(class(data[,i]),
					"numeric"=0,
					"factor"=1)
			}
			if(is.null(family)) family=fam
			data <- data
		}
	}
		
	mod <- mix(response=form,data=data,nstates=nclasses,family=family, ...)
	attr(mod,"type") <- "lca"
	if(fit) res <- fit(mod, emcontrol=em.control(maxit=500), verbose=verbose)
	else res <- mod
	return(res)
}

hmm <- function(data, nstates, fit=TRUE, ntimes=NULL, family=NULL, verbose=TRUE, ...) {
	
	# univariate numeric data by default treated as gaussian
 	if(is.vector(data)) {
		nm <- deparse(substitute(data))
		form <- as.formula(paste(nm,"~1",sep=""))
		if(is.null(family)) family=gaussian()
		if(is.null(ntimes)) ntimes <- length(data)
		simple <- 0
		data <- NULL
	} else {
		# univariate factor data treated as multinomial
 		if(is.factor(data)) {
			nm <- deparse(substitute(data))
			form <- as.formula(paste(nm,"~1",sep=""))
			if(is.null(ntimes)) ntimes <- length(data)
			if(is.null(family)) family=multinomial("identity")
			simple <- 1
			data <- NULL
		} else {
			# multivariate data
			form <- list()
			fam <- list()
			nc <- ifelse(!is.null(dim(data)), dim(data)[2],1)
			simple <- numeric(nc)
			for(i in 1:nc) {
				form[[i]] <- as.formula(paste(names(data)[i],"~1",sep=""))
				fam[[i]] <- switch(class(data[,i]),
					"numeric"=gaussian(),
					"factor"=multinomial("identity"),
					stop("Provide family arguments of data other than 'numeric' or 
						'factor's"))
				simple[i] <- switch(class(data[,i]),
					"numeric"=0,
					"factor"=1)
			}
			if(is.null(family)) family=fam
			if(is.null(ntimes)) ntimes=nrow(data)
			data <- data
		}
	}
		
	mod <- depmix(response=form,data=data,nstates=nstates,ntimes=ntimes,family=family)
	attr(mod,"type") <- "hmm"
	if(fit) res <- fit(mod, emcontrol=em.control(maxit=500), verbose=verbose)
	else res <- mod
	return(res)
}




R/long2d2.R

setwd("~/Documents/Dropbox/hmmr/datasets")

dt <- read.spss("long.sav",to.data.frame=TRUE)

attach(dt)

sel <- c("NUMMER","GESLACHT","GEBOORTE", "SCHOOL",   "GROEP",    
	"GROEPSL", "DAYS",     "YEARS",    
	"AANWTM1", "AANWTM2",  "AANWTM3",   "AANWTM4" ,
	"AANWTM5",  "AANWTM6",  "AANWTM7", "AANWTM8", 
	"REGELS1",  "REGELS2", "REGELS3",  "REGELS4",  "REGELS5",  "REGELS6",  
	"REGELS7",  "REGELS8", 
	"SCLLEVEL")


# select relevant columns
d1 <- dt[,sel]

d2 <- reshape(d1,varying=list(c("AANWTM1", "AANWTM2",  "AANWTM3",   "AANWTM4" ,
	"AANWTM5",  "AANWTM6",  "AANWTM7", "AANWTM8"),
	c("REGELS1",  "REGELS2", "REGELS3",  "REGELS4",  "REGELS5",  "REGELS6",  
	"REGELS7",  "REGELS8")),sep="",direction="long")

# compute sum scores of different item types

afn <- paste("AFN",1:8,sep="")

# weight items
afng <- paste(afn,c("_G"),sep="")
afng <- sapply(afng,paste,1:5,sep="")

# afname 1
gc <- apply(dt[,afng[,1]],1,sum,na.rm=TRUE)
gi <- apply(1-dt[,afng[,1]],1,sum,na.rm=TRUE)

# afname 2-8
for(i in 2:8) {
	gc <- c(gc,apply(dt[,afng[,i]],1,sum,na.rm=TRUE))
	gi <- c(gi,apply(1-dt[,afng[,i]],1,sum,na.rm=TRUE))
}

gs <- gc + gi

d2$"gc" <- gc
d2$"gi" <- gi
d2$"gs" <- gs



# distance items
afna <- paste(afn,c("_A"),sep="")
afna <- sapply(afna,paste,1:5,sep="")

# afname 1
ac <- apply(dt[,afna[,1]],1,sum,na.rm=TRUE)
ai <- apply(1-dt[,afna[,1]],1,sum,na.rm=TRUE)

# afname 2-8
for(i in 2:8) {
	ac <- c(ac,apply(dt[,afna[,i]],1,sum,na.rm=TRUE))
	ai <- c(ai,apply(1-dt[,afna[,i]],1,sum,na.rm=TRUE))
}

as <- ac + ai

d2$"ac" <- ac
d2$"ai" <- ai
d2$"as" <- as



# conflict weight items
afncg <- paste(afn,c("_CG"),sep="")
afncg <- sapply(afncg,paste,1:5,sep="")

# afname 1
cgc <- apply(dt[,afncg[,1]],1,sum,na.rm=TRUE)
cgi <- apply(1-dt[,afncg[,1]],1,sum,na.rm=TRUE)

# afname 2-8
for(i in 2:8) {
	cgc <- c(cgc,apply(dt[,afncg[,i]],1,sum,na.rm=TRUE))
	cgi <- c(cgi,apply(1-dt[,afncg[,i]],1,sum,na.rm=TRUE))
}

cgs <- cgc + cgi

d2$"cgc" <- cgc
d2$"cgi" <- cgi
d2$"cgs" <- cgs



# conflict distance
afnca <- paste(afn,c("_CA"),sep="")
afnca <- sapply(afnca,paste,1:5,sep="")

# afname 1
cac <- apply(dt[,afnca[,1]],1,sum,na.rm=TRUE)
cai <- apply(1-dt[,afnca[,1]],1,sum,na.rm=TRUE)

# afname 2-8
for(i in 2:8) {
	cac <- c(cac,apply(dt[,afnca[,i]],1,sum,na.rm=TRUE))
	cai <- c(cai,apply(1-dt[,afnca[,i]],1,sum,na.rm=TRUE))
}

cas <- cac + cai

d2$"cac" <- cac
d2$"cai" <- cai
d2$"cas" <- cas



# conflict balance
afncb <- paste(afn,c("_CB"),sep="")
afncb <- sapply(afncb,paste,1:5,sep="")

# afname 1
cbc <- apply(dt[,afncb[,1]],1,sum,na.rm=TRUE)
cbi <- apply(1-dt[,afncb[,1]],1,sum,na.rm=TRUE)

# afname 2-8
for(i in 2:8) {
	cbc <- c(cbc,apply(dt[,afncb[,i]],1,sum,na.rm=TRUE))
	cbi <- c(cbi,apply(1-dt[,afncb[,i]],1,sum,na.rm=TRUE))
}

cbs <- cbc + cbi

d2$"cbc" <- cbc
d2$"cbi" <- cbi
d2$"cbs" <- cbs


# score op oefenitems

afnoef <- paste(afn,c("_"),sep="")
afnoef <- sapply(afnoef,paste,1:3,sep="")
afnoef <- matrix(paste(afnoef,c("C"),sep=""),3,8)

# afname 1
oef <- apply(dt[,afnoef[,1]],1,sum,na.rm=TRUE)

# afname 2-8
for(i in 2:8) {
	oef <- c(oef,apply(dt[,afnoef[,i]],1,sum,na.rm=TRUE))
}

d2$"oef" <- oef


save(d2, file="d2.Rda")






R/stationTest.R

library(depmixS4)

setwd("~/Documents/projects/depmixProject/codesvn")

setwd("/Users/ivisser1/Documents/Dropbox/hmmfunction")

options(digits=4)


data(speed)
speed$corr <- as.factor(speed$corr)

data(balance)
balance <- balance[order(balance$agedays),]
bal <- balance[,4:6]
bal$t1 <- as.factor(bal$t1)
bal$t2 <- as.factor(bal$t2)
bal$t3 <- as.factor(bal$t3)

source("hmm.R")

# four binary items on the balance scale task
mod4 <- mix(list(t1~1,t2~1,t3~1), data=balance, nstates=2,
	family=list(multinomial("identity"),multinomial("identity"),multinomial("identity")))

fm4 <- fit(mod4)

summary(fm4)

m3 <- lca(bal,nc=2)

m3 <- lca(bal[,1:2],nc=2)



summary(m3,com=T)

h1 <- hmm(speed[,1],ns=2)

h1

summary(h1,com=T)

h2 <- hmm(speed[,1],ns=2)

summary(h2,com=T)


h3 <- hmm(speed$rt,ns=2)
summary(h3,com=T)


h4 <- hmm(speed[,1:2],ns=2)
summary(h4,com=T)


m2 <- lca(speed[,1:2],nc=2)








R/hmmBasisSchool.R
setwd("~/Documents/Dropbox/hmmr/datasets")

load(file="d2.Rda")

# reorder to arrange data per subject
d2 <- d2[order(d2$"NUMMER"),]

# select complete data
sel <- d2$NUMMER[which(d2$"AANWTM1"=="Selected"&d2$time==8)]
d2 <- d2[d2$NUMMER%in%sel,]

sel2 <- "basis"

d2 <- d2[d2$SCLLEVEL%in%sel2,]

pps <- unique(d2$NUMMER)[which(by(d2$oef,list(d2$NUMMER),sum)>22)]

d2 <- d2[d2$SCLLEVEL%in%pps,]


# onderstaand geeft een mooi model met R0, R1, R2, R3/ADD
# meer startwaarden nodig, beste likelihood tot nu toe: -7788.696
# (voor het 4-state model)

hm4 <- depmix(list(cbind(gc,gi)~1,cbind(ac,ai)~1,cbind(cgc,cgi)~1), 
	data=d2, family=list(binomial(),binomial(),binomial()), ntimes=rep(8,470), ns=4)
fhm4 <- fit(hm4,emcontrol=em.control(maxit=400))


# beste ll tot nu toe: -7364
hm5 <- depmix(list(cbind(gc,gi)~1,cbind(ac,ai)~1,cbind(cgc,cgi)~1), 
	data=d2, family=list(binomial(),binomial(),binomial()), 
	ntimes=rep(8,470), ns=5)
fhm5 <- fit(hm5,emcontrol=em.control(maxit=400))



hm6 <- depmix(list(cbind(gc,gi)~1,cbind(ac,ai)~1,cbind(cgc,cgi)~1), 
	data=d2, family=list(binomial(),binomial(),binomial()), 
	ntimes=rep(8,470), ns=6)
fhm6 <- fit(hm6,emcontrol=em.control(maxit=400))

hm7 <- depmix(list(cbind(gc,gi)~1,cbind(ac,ai)~1,cbind(cgc,cgi)~1), 
	data=d2, family=list(binomial(),binomial(),binomial()), 
	ntimes=rep(8,470), ns=7)
fhm7 <- fit(hm7,emcontrol=em.control(maxit=400))


R/analyses.R

setwd("~/Documents/Dropbox/hmmr/datasets")

load(file="d2.Rda")

# reorder to arrange data per subject
d2 <- d2[order(d2$"NUMMER"),]

# select complete data
sel <- d2$NUMMER[which(d2$"AANWTM1"=="Selected"&d2$time==8)]
d2 <- d2[d2$NUMMER%in%sel,]

sel2 <- "basis"

d2 <- d2[d2$SCLLEVEL%in%sel2,]

pps <- unique(d2$NUMMER)[which(by(d2$oef,list(d2$NUMMER),sum)>22)]

d2 <- d2[d2$SCLLEVEL%in%pps,]




hm4 <- depmix(list(cbind(gc,gi)~1,cbind(ac,ai)~1,cbind(cgc,cgi)~1), 
	data=d2, family=list(binomial(),binomial(),binomial()), ntimes=rep(8,470), ns=4)
fhm4 <- fit(hm4,emcontrol=em.control(maxit=400))




# do some models ...
library(depmixS4)


# lca models on weight and distance items only
m1 <- mix(list(cbind(gc,gi)~1,cbind(ac,ai)~1),data=d2, family=list(binomial(),binomial()),ns=1)
fm1 <- fit(m1)

m2 <- mix(list(cbind(gc,gi)~1,cbind(ac,ai)~1),data=d2, family=list(binomial(),binomial()),ns=2)
fm2 <- fit(m2)


fm3 <- list()
m3 <- mix(list(cbind(gc,gi)~1,cbind(ac,ai)~1),data=d2, family=list(binomial(),binomial()),ns=3)
for(i in 1:10) {
	fm3[[i]] <- fit(m3)
}


fm3 <- fm3[[1]]

m4 <- mix(list(cbind(gc,gi)~1,cbind(ac,ai)~1),data=d2, family=list(binomial(),binomial()),ns=4)
fm4 <- fit(m4,emcontrol=em.control(maxit=400))
# fm4 <- fit(fm4)

m5 <- mix(list(cbind(gc,gi)~1,cbind(ac,ai)~1),data=d2, family=list(binomial(),binomial()),ns=5)
fm5 <- fit(m5,emcontrol=em.control(maxit=400))

fm5 <- fit(fm5,emcontrol=em.control(maxit=400,random.start=FALSE))

save(fm1,fm2,fm3,fm4,fm5,file="GplusAmodels.Rda")


# lca models on weigth, distance, and conflict weight

# lca models on weight and distance items only
m1 <- mix(list(cbind(gc,gi)~1,cbind(ac,ai)~1,cbind(cgc,cgi)~1),
	data=d2, family=list(binomial(),binomial(),binomial()),ns=1)
fm1 <- fit(m1)

m1 <- mix(list(cbind(gc,gi)~1,cbind(ac,ai)~1,cbind(cgc,cgi)~1),
	data=d2, family=list(binomial(),binomial(),binomial()),ns=2)
fm2 <- fit(m1)

m1 <- mix(list(cbind(gc,gi)~1,cbind(ac,ai)~1,cbind(cgc,cgi)~1),
	data=d2, family=list(binomial(),binomial(),binomial()),ns=3)
fm3 <- fit(m1)

m1 <- mix(list(cbind(gc,gi)~1,cbind(ac,ai)~1,cbind(cgc,cgi)~1),
	data=d2, family=list(binomial(),binomial(),binomial()),ns=4)
fm4 <- fit(m1)

m1 <- mix(list(cbind(gc,gi)~1,cbind(ac,ai)~1,cbind(cgc,cgi)~1),
	data=d2, family=list(binomial(),binomial(),binomial()),ns=5)
fm5 <- fit(m1)

m1 <- mix(list(cbind(gc,gi)~1,cbind(ac,ai)~1,cbind(cgc,cgi)~1),
	data=d2, family=list(binomial(),binomial(),binomial()),ns=6)
fm6 <- fit(m1)




# # use multinomial ipv binomial
# mm2 <- mix(list(cbind(gc,gi)~1,cbind(ac,ai)~1),
# 	data=d2, family=list(multinomial("identity"),multinomial("identity")), ns=2)
# 
# fmm2 <- fit(mm2)



# lca models on all items

ma1 <- mix(list(cbind(gc,gi)~1,cbind(ac,ai)~1,cbind(cgc,cgi)~1,cbind(cac,cai)~1,cbind(cbc,cbi)~1), 
	data=d2, family=list(binomial(),binomial(),binomial(),binomial(),binomial()), ns=1)
fma1 <- fit(ma1)


ma2 <- mix(list(cbind(gc,gi)~1,cbind(ac,ai)~1,cbind(cgc,cgi)~1,cbind(cac,cai)~1,cbind(cbc,cbi)~1), 
	data=d2, family=list(binomial(),binomial(),binomial(),binomial(),binomial()), ns=2)
fma2 <- fit(ma2,emcontrol=em.control(maxit=400))


ma3 <- mix(list(cbind(gc,gi)~1,cbind(ac,ai)~1,cbind(cgc,cgi)~1,cbind(cac,cai)~1,cbind(cbc,cbi)~1), 
	data=d2, family=list(binomial(),binomial(),binomial(),binomial(),binomial()), ns=3)
fma3 <- fit(ma3,emcontrol=em.control(maxit=400))


ma4 <- mix(list(cbind(gc,gi)~1,cbind(ac,ai)~1,cbind(cgc,cgi)~1,cbind(cac,cai)~1,cbind(cbc,cbi)~1), 
	data=d2, family=list(binomial(),binomial(),binomial(),binomial(),binomial()), ns=4)
fma4 <- fit(ma4,emcontrol=em.control(maxit=400))


ma5 <- mix(list(cbind(gc,gi)~1,cbind(ac,ai)~1,cbind(cgc,cgi)~1,cbind(cac,cai)~1,cbind(cbc,cbi)~1), 
	data=d2, family=list(binomial(),binomial(),binomial(),binomial(),binomial()), ns=5)
fma5 <- fit(ma5,emcontrol=em.control(maxit=400))


ma6 <- mix(list(cbind(gc,gi)~1,cbind(ac,ai)~1,cbind(cgc,cgi)~1,cbind(cac,cai)~1,cbind(cbc,cbi)~1), 
	data=d2, family=list(binomial(),binomial(),binomial(),binomial(),binomial()), ns=6)
fma6 <- fit(ma6,emcontrol=em.control(maxit=400))


ma7 <- mix(list(cbind(gc,gi)~1,cbind(ac,ai)~1,cbind(cgc,cgi)~1,cbind(cac,cai)~1,cbind(cbc,cbi)~1), 
	data=d2, family=list(binomial(),binomial(),binomial(),binomial(),binomial()), ns=7)
fma7 <- fit(ma7,emcontrol=em.control(maxit=400))


ma8 <- mix(list(cbind(gc,gi)~1,cbind(ac,ai)~1,cbind(cgc,cgi)~1,cbind(cac,cai)~1,cbind(cbc,cbi)~1), 
	data=d2, family=list(binomial(),binomial(),binomial(),binomial(),binomial()), ns=8)
fma8 <- fit(ma8,emcontrol=em.control(maxit=400))


ma9 <- mix(list(cbind(gc,gi)~1,cbind(ac,ai)~1,cbind(cgc,cgi)~1,cbind(cac,cai)~1,cbind(cbc,cbi)~1), 
	data=d2, family=list(binomial(),binomial(),binomial(),binomial(),binomial()), ns=9)
fma9 <- fit(ma9,emcontrol=em.control(maxit=400))


save(fma1,fma2,fma3,fma4,fma5,fma6,fma7,fma8,file="lcaAllItems.Rda")


# hmm models on all items

hm1 <- depmix(list(cbind(gc,gi)~1,cbind(ac,ai)~1,cbind(cgc,cgi)~1,cbind(cac,cai)~1,cbind(cbc,cbi)~1), 
	data=d2, family=list(binomial(),binomial(),binomial(),binomial(),binomial()), ns=1, ntimes=rep(8,608))
fhm1 <- fit(hm1,emcontrol=em.control(maxit=400))


hm2 <- depmix(list(cbind(gc,gi)~1,cbind(ac,ai)~1,cbind(cgc,cgi)~1,cbind(cac,cai)~1,cbind(cbc,cbi)~1), 
	data=d2, family=list(binomial(),binomial(),binomial(),binomial(),binomial()), ntimes=rep(8,608), ns=2)
fhm2 <- fit(hm2,emcontrol=em.control(maxit=400))



hm5 <- depmix(list(cbind(gc,gi)~1,cbind(ac,ai)~1,cbind(cgc,cgi)~1,cbind(cac,cai)~1,cbind(cbc,cbi)~1), 
	data=d2, family=list(binomial(),binomial(),binomial(),binomial(),binomial()), ntimes=rep(8,608), ns=5)
fhm5 <- fit(hm5,emcontrol=em.control(maxit=400))


hm6 <- depmix(list(cbind(gc,gi)~1,cbind(ac,ai)~1,cbind(cgc,cgi)~1,cbind(cac,cai)~1,cbind(cbc,cbi)~1), 
	data=d2, family=list(binomial(),binomial(),binomial(),binomial(),binomial()), ntimes=rep(8,608), ns=6)
fhm6 <- fit(hm6,emcontrol=em.control(maxit=400))


hm7 <- depmix(list(cbind(gc,gi)~1,cbind(ac,ai)~1,cbind(cgc,cgi)~1,cbind(cac,cai)~1,cbind(cbc,cbi)~1), 
	data=d2, family=list(binomial(),binomial(),binomial(),binomial(),binomial()), ntimes=rep(8,608), ns=7)
fhm7 <- fit(hm7,emcontrol=em.control(maxit=400))


save(fhm5,fhm6, file="hmmAllItems.Rda")


